---
title: "Mitochondrial ubiquitin landscape in neurons"
author: "Marek Gierlinski"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    css: tabbed_notebook.css
---

Collaborators: Miratul Muqit

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  warning = FALSE,
  message = FALSE,
  echo = FALSE,
  cache = TRUE
)
library(drake)
library(kableExtra)
library(tidyverse)

myKable <- function(df, row.names=FALSE, col.names=NA, digits=2, bootstrap="condensed", font.size=12, label=NULL, ...) {
  kable(df, format="html", row.names=row.names, col.names=col.names, digits=digits, label=label, ...) %>% kable_styling(bootstrap_options=bootstrap, full_width=FALSE, position="left", font_size=font.size)
}
N <- function(n) prettyNum(n, big.mark = ",")

public_html <- "http://www.compbio.dundee.ac.uk/user/mgierlinski/ubimito"
```

# {.tabset}

## Data

Data came from quantitative diGlycine (diGly) capture proteomics to identify Parkin-dependent ubiquitylation. The experiment was carried out in two conditions, unstimulated/basal (UT) and stimulated/depolarised (AO) in five replicates each. The total proteome was also analysed.

There are `r N(readd(kgg_n_sites))` sites on `r N(readd(kgg_n_prot))` individual proteins in the diGly data. There are `r N(readd(total_n_prot))` proteins in the total proteome data.

## Databases

### MitoCarta

We use [MitoCarta 2.0](https://www.broadinstitute.org/files/shared/metabolism/mitocarta/human.mitocarta2.0.html) to identify mitochondrial proteins. This database contains `r N(readd(n_mito))` genes.

### E2 enzymes and E3 ligases

We identify E2 and E3 using [UbiHub](https://ubihub.thesgc.org/static/UbiHub.html). We have the following numbers of enzymes and ligases in this database:

```{r ubihub}
readd(ubihub) %>% group_by(ubi_part) %>% tally() %>% rename(type = ubi_part) %>% myKable()
```


## Results

```{r ubi_tot_mito}
loadd(kgg_mito, tot_mito)

```

### Differential expression

Proteomics data came with Welch's t-test p-values and fold changes. I am not sure if these are corrected for multiple tests.

Here, we re-do differential expression analysis using a more sophisticated tool, `limmma` and making sure that p-values are corrected for multiple tests. `limma` borrows information across features (sites) in order to moderate variance. We also normalise each ratio to its median. The raw data medians are close, but not equal to one, so this step reduces variance between replicates slightly and improves power of the test.

Result is not very different from the original Welch's tests. Without the normalisation step, the numbers of significant sites (diGly) is very similar. With normalisation and using `limma`, we get `r N(readd(n_kgg_de_fdr))` statistically significant sites. The Welch's p-value gives `r N(readd(n_kgg_de_welch))` DE sites. The difference is mostly due to normalisation step.

This figure shows a volcano plot. Black points indicate data present in MitoCarta. The dashed horizontal line shows the significance limit, FDR = 0.05.

```{r kgg_volcano, fig.width=4, fig.height=4}
readd(fig_kgg_volcano)
kgg_de_link <- file.path(public_html, "tab/kgg_de.tsv")
```


[Download data](`r kgg_de_link`)

### Identity of ubiquitylated proteins

Since we do not have sub-compartment data yet (it will come with MitoCarta 3.0), we can only identify which ubiquitilated proteins are in the MitoCarta database. However, these are already identified in the Excel file.

As we want proteins in basal state, we select only diGly proteins that are detected in the total proteome. This gives us `r N(readd(kgg_n_prot_basal))` proteins. Out of these, there are `r N(readd(n_kgg_in_mito))` diGly proteins in MitoCarta database. None of them is E2 or E3.

```{r datatable_settings, message=FALSE, warning=FALSE}
dummy <- DT:::DT2BSClass(c('compact', 'cell-border'))
```


```{r ubi_mito, cache=FALSE}
kgg_mito %>%
  filter(in_mito) %>% 
  select(uniprot, gene_name, site_position, description, p_value, fdr, log_fc) %>% 
  #myKable() %>% 
  #scroll_box(width="100%", height="300px")
  DT::datatable() %>% 
  DT::formatSignif(c("p_value", "fdr", "log_fc"), 2) %>% 
  DT::formatStyle(columns = colnames(.$x$data), fontSize = '80%')

kgg_mito_link <- file.path(public_html, "tab/kgg_mito.tsv")
```

[Download data](`r kgg_mito_link`)

### Whole proteome E2/E3 

We identify following numbers of E2/E3 inside and outside of mitochondria:

```{r tot_not_mito_nums}
tot_mito %>% group_by(ubi_part, in_mito) %>% tally() %>% drop_na() %>% arrange(in_mito) %>% myKable()
```

Here is a full table of these proteins.

```{r tot_mito_tab, cache=FALSE}
tot_mito %>% 
  select(-id) %>% 
  filter(!is.na(ubi_part)) %>% 
  DT::datatable() %>% 
  DT::formatSignif(c("p_value", "fdr", "log_fc"), 2) %>% 
  DT::formatStyle(columns = colnames(.$x$data), fontSize = '80%')
  #myKable() %>% 
  #scroll_box(width="100%", height="300px")

total_mito_link <- file.path(public_html, "tab/total_mito.tsv")
```

[Download data](`r total_mito_link`)

### Changing sites

The next figure shows how individual sites in each protein change upon stimulation. Only diGly proteins present in MitoCarta are shown. Ubiquitylation sites are simply indexed 1, 2, 3,... from the smallest to the largest position in the protein. White fill indicates non-significant change (FDR > 0.05).

```{r mito_change, fig.width=5, fig.height=16}
readd(fig_mito_change) 
```
